docs/predictive-analytics/model-retraining/process.sql

```diff
@@ -0,0 +1,66 @@
+SET 'auto.offset.reset' = 'earliest';
+
+-- Create stream of predictions
+CREATE STREAM PREDICTED_WEIGHT(
+  "Fish_Id" VARCHAR KEY,
+  "Species" VARCHAR,
+  "Height" DOUBLE,
+  "Length" DOUBLE,
+  "Prediction" DOUBLE
+  )
+WITH(
+  KAFKA_TOPIC = 'kt.mdb.weight-prediction', 
+  VALUE_FORMAT = 'JSON',
+  PARTITIONS = 6
+);
+
+-- Create stream of actual weights
+CREATE STREAM ACTUAL_WEIGHT(
+  "Fish_Id" VARCHAR KEY,
+  "Species" VARCHAR,
+  "Weight" DOUBLE
+  )
+WITH(
+  KAFKA_TOPIC = 'kt.mdb.machine-weight', 
+  VALUE_FORMAT = 'JSON',
+  PARTITIONS = 6
+);
+
+-- Create stream joining predictions with actual weights
+CREATE STREAM DIFF_WEIGHT
+WITH(
+  KAFKA_TOPIC = 'diff_weight', 
+  VALUE_FORMAT = 'JSON'
+)
+AS SELECT
+   'Key' AS "Key",
+   PREDICTED_WEIGHT."Fish_Id" AS "Fish_Id",
+   PREDICTED_WEIGHT."Species" AS "Species",
+   PREDICTED_WEIGHT."Length" AS "Length",
+   PREDICTED_WEIGHT."Height" AS "Height",
+   PREDICTED_WEIGHT."Prediction" AS "PredictedWeight",
+   ACTUAL_WEIGHT."Weight" AS "ActualWeight",
+   ROUND(ABS(PREDICTED_WEIGHT."Prediction" - ACTUAL_WEIGHT."Weight") / ACTUAL_WEIGHT."Weight", 3) AS "Error"
+FROM PREDICTED_WEIGHT
+INNER JOIN ACTUAL_WEIGHT
+WITHIN 1 MINUTE
+ON PREDICTED_WEIGHT."Fish_Id" = ACTUAL_WEIGHT."Fish_Id";
+
+-- Create table of one minute aggregates with over 15% error rate
+set 'ksql.suppress.enabled'='true';
+CREATE TABLE RETRAIN_WEIGHT
+WITH(
+  KAFKA_TOPIC = 'retrain_weight', 
+  VALUE_FORMAT = 'JSON'
+)
+AS SELECT
+   "Key",
+   COLLECT_SET("Species") AS "Species",
+   EARLIEST_BY_OFFSET("Fish_Id") AS "Fish_Id_Start",
+   LATEST_BY_OFFSET("Fish_Id") AS "Fish_Id_End",
```

Comment:
Not sure if I understand the purpose of `EARLIEST_BY_OFFSET` and `LATEST_BY_OFFSET`.

Reply:
This provides the start and end of the set of data that will be used by the model retraining process. This one record will be used to trigger the retraining process but the process itself will need all of the records that went into creating that aggregate record.

Reply:
Given that we need to repartition the data on the surrogate `'key'` attribute, it seems that there is no ordering guarantee on the internal input topic (ie, repartition topic) of the aggregation. Thus, it seems we effectively pick two random `Fish_Id` -- and there is no guarantee what the Id in-between are.

> This provides the start and end of the set of data that will be used by the model retraining process.

It seems to picks the fish-id of the first record in offset order in the repartition topic and the fish-id of the last record in offset order in the repartition topic. But as we have interleaved write into the repartition topic, it seems to be possible to actually reason about it (not will be be deterministic)?

If the behavior is by-design of the recipe I am happy with it. Just want to make sure we are on the same page. I don't know the use-case background nor blog post, so maybe I don't have enough context information: from my point it just seems not semantically meaningful?

Reply:
I understand what you're saying, @mjsax, but it does seem to work.  I've run multiple sets of data through and it does group them as expected and it does give me the low and high Fish_Id values.  This is with 6 partitions in the source topics.

Reply:
Interesting. How much data did you process? Also, if you run the same query over the same input data multiple times, does it always compute the exact same result?

I leave it up to your own judgement in the end -- just curious :)

Reply:
We do get indeterminate results when running this query over existing data but it seems to always work on new incoming data, and that is the goal of the use case.